### DISCUSSION TITLE
Transition into a Docker container

### DISCUSSION THREAD

**Discussion by AndersSundman** [bazelbuild/bazel#21676] [Design proposals]
When building embedded code, bazel toolchain transitions are very useful. We build a PC application with one toolchain and transition to an embedded toolchain to build the firmware that we then embed into the PC application as a firmware upgrade blob before the build is over. Nice!

BUT... what if you can't run all your toolchains on the same machine? Some embedded toolchains only work on Windows, others only work on Linux. It would be awesome if we could start building on Windows and when we need to transition to the Linux embedded toolchain, bazel can not only transition the toolchain but also jump into a Linux docker container to use it.

**Comment by veith4f** [comment]
Great question. I am still relatively new to bazel but definitely wonder if in the heavily containerized world we have today there is still so much sense for platform-specific toolchains and rule implementations that ship different binaries for different platforms.

I just took over maintenance for a large bazel build env that going forward should be usable on different build machines comprising windows, linux and macos/amd64, macos/arm. 

Quite frankly, I think it is kind of ridiculous to implement treatments for all these different platforms. Bazel should run in a container where the respective bazel version is installed and bazel rules should make heavy use of bash and other gnu tools that the container provides. 

There should be a `bash` rule that allows running bash scripts and a large lib of rules that build on that `bash` rule to implement stuff like `base64-encode/decode`, `yq`, `jq`, `sed`, and all these other targets that sort of naturally suggest themselves.

I really wonder why that doesn't exist yet. Wouldn't it make much more sense to use bazel that way?

**Comment by veith4f** [comment]
Proposal: https://github.com/bazelbuild/bazel/discussions/26891
